File: A2Z_striver/1 Basics/1.5 Hashing/sum_hghest_lowest_frq/sum_hghest_lowest_frq.py
from collections import Counter
import sys

# Header Comment Block
# Problem: Sum of Highest and Lowest Frequencies.
# Input: Array of integers.
# Output: Sum (int).
# Goal: O(N) Time.

class Solution:

    # ~ Approach 1: Sorting (Your Solution)
    # Analysis: Time O(N + U log U) | Space O(N)
    # Note: 'U' is the number of unique elements. 
    # If all elements are unique, this becomes O(N log N).
    def sumHighestAndLowestSorting(self, nums):
        if not nums: return 0
        
        count = Counter(nums)
        
        # - Step 1: Get Unique Sorted Frequencies
        # set() removes duplicates, sorted() orders them
        Frq = sorted(list(set(count.values())))
        
        # - Step 2: Add First (Min) and Last (Max)
        return Frq[0] + Frq[-1]


    # ~ Approach 2: Min/Max Functions (Optimal / Pythonic)
    # Analysis: Time O(N) | Space O(N)
    # Why: We avoid sorting entirely. Finding min and max takes O(U) time.
    def sumHighestAndLowestOptimal(self, nums):
        if not nums: return 0
        
        # - Step 1: Count Frequencies
        counts = Counter(nums)
        
        # - Step 2: Get all frequency values
        freq_values = counts.values()
        
        # - Step 3: Find Max and Min directly
        # Internal Logic: Iterates through the list once for max, once for min.
        high_freq = max(freq_values)
        low_freq  = min(freq_values)
        
        return high_freq + low_freq

    # ~ Approach 3: Manual Loop (The "Under the Hood" Way)
    # Analysis: Time O(N) | Space O(N)
    def sumHighestAndLowestManual(self, nums):
        if not nums: return 0
        
        # - Step 1: Count Frequencies
        # Time: O(N)
        counts = Counter(nums)
        
        # - Step 2: Initialize Extremes
        # We start min with infinity so the first number we see is smaller.
        # We start max with -1 (or -inf) so the first number is larger.
        min_freq = float('inf')
        max_freq = float('-inf')
        
        # - Step 3: Single Pass Loop
        # Time: O(U) where U is unique elements
        for freq in counts.values():
            
            # Update Max
            if freq > max_freq:
                max_freq = freq
                
            # Update Min
            if freq < min_freq:
                min_freq = freq
                
        # - Step 4: Return Sum
        return max_freq + min_freq


# --- Driver Code ---
if __name__ == "__main__":
    sol = Solution()
    
    # Ex 1: Freqs {1:1, 2:2, 3:3} -> Min 1, Max 3 -> Sum 4
    arr1 = [1, 2, 2, 3, 3, 3]
    
    # Ex 2: Freqs {4:2, 5:2, 6:1} -> Min 1, Max 2 -> Sum 3
    arr2 = [4, 4, 5, 5, 6]
    
    # Edge Case: Single Element [1] -> Freq {1:1} -> Min 1, Max 1 -> Sum 2
    arr3 = [1]

    print(f"Sorting Approach (Ex1): {sol.sumHighestAndLowestSorting(arr1)}")
    print(f"Optimal Approach (Ex1): {sol.sumHighestAndLowestOptimal(arr1)}")
    print(f"Optimal Approach (Ex3): {sol.sumHighestAndLowestOptimal(arr3)}")

File: A2Z_striver/1 Basics/1.5 Hashing/sum_hghest_lowest_frq/sum_hghest_lowest_frq.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm> // for sort, min, max
#include <climits>   // for INT_MAX, INT_MIN

using namespace std;

// Header Comment Block
// Problem: Sum of Highest and Lowest Frequencies.
// Input: Vector of integers.
// Output: Sum (int).
// Goal: O(N) Time | O(N) Space.

class Solution {
public:

    // ~ Approach 1: Sorting (Naive)
    // Analysis: Time O(N + U log U) | Space O(N)
    // Logic: Dump frequencies into a vector, sort it, pick ends.
    int sumHighLowSorting(vector<int>& arr) {
        if (arr.empty()) return 0;

        // - Step 1: Count Frequencies
        unordered_map<int, int> freqMap;
        for (int x : arr) {
            freqMap[x]++;
        }

        // - Step 2: Extract Frequencies into a Vector
        vector<int> frequencies;
        for (auto it : freqMap) {
            frequencies.push_back(it.second);
        }

        // - Step 3: Sort
        sort(frequencies.begin(), frequencies.end());

        // - Step 4: Sum First (Min) and Last (Max)
        return frequencies.front() + frequencies.back();
    }


    // ~ Approach 2: STL Algorithms (Clean One-Pass)
    // Analysis: Time O(N) | Space O(N)
    // Logic: Use std::min and std::max inside the loop implicitly.
    int sumHighLowSTL(vector<int>& arr) {
        if (arr.empty()) return 0;

        unordered_map<int, int> freqMap;
        for (int x : arr) freqMap[x]++;

        int maxFreq = INT_MIN;
        int minFreq = INT_MAX;

        for (auto it : freqMap) {
            // it->second holds the frequency
            maxFreq = max(maxFreq, it.second);
            minFreq = min(minFreq, it.second);
        }

        return maxFreq + minFreq;
    }


    // ~ Approach 3: Manual Loop (The "Under the Hood" Way)
    // Analysis: Time O(N) | Space O(N)
    // Logic: Pure conditional checks, no helper functions.
    int sumHighLowManual(vector<int>& arr) {
        if (arr.empty()) return 0;

        // - Step 1: Count Frequencies
        unordered_map<int, int> freqMap;
        for (int x : arr) {
            freqMap[x]++;
        }

        // - Step 2: Initialize Extremes
        // Set min to MAX_INT so the first value is smaller
        // Set max to MIN_INT so the first value is larger
        int minFreq = INT_MAX;
        int maxFreq = INT_MIN;

        // - Step 3: Single Pass Loop
        // Iterate over the map entries
        for (auto it : freqMap) {
            int currentFreq = it.second;

            // Update Max
            if (currentFreq > maxFreq) {
                maxFreq = currentFreq;
            }

            // Update Min
            if (currentFreq < minFreq) {
                minFreq = currentFreq;
            }
        }

        return maxFreq + minFreq;
    }
};

int main() {
    // Fast I/O
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    Solution sol;

    // Test Case 1: [1, 2, 2, 3, 3, 3]
    // Frequencies: {1:1, 2:2, 3:3} -> Min 1, Max 3 -> Sum 4
    vector<int> arr1 = {1, 2, 2, 3, 3, 3};

    // Test Case 2: [4, 4, 5, 5, 6]
    // Frequencies: {4:2, 5:2, 6:1} -> Min 1, Max 2 -> Sum 3
    vector<int> arr2 = {4, 4, 5, 5, 6};

    cout << "Sorting Approach (Arr1): " << sol.sumHighLowSorting(arr1) << endl;
    cout << "STL Approach (Arr1):     " << sol.sumHighLowSTL(arr1) << endl;
    cout << "Manual Approach (Arr2):  " << sol.sumHighLowManual(arr2) << endl;

    return 0;
}